File: leetcode/leetcode_46.py
# DFS를 이용한 풀이
class Solution:
    def permute(self, nums : int):
        answer = []
        visited = set()

        def dfs(idx, curr = []):
            if len(curr) == len(nums):
                answer.append(curr[:])
                return
            
            for i in range(len(nums)):
                if i not in visited:
                    visited.add(i)
                    curr.append(nums[i])
                    dfs(i, curr)
                    curr.pop()
                    visited.remove(i)
        dfs(0)
        return answer
        

# itertools 라이브러리를 이용한 풀이
import itertools

class Solution:
    def permute(self, nums):
        return list(itertools.permutations(nums))

File: leetcode/leetcode_46.js
const permute = (nums) => {
  const answer = [];
  const visited = new Set();

  const dfs = (idx, curr = []) => {
    if (curr.length == nums.length) {
      answer.push(curr.slice());
      return;
    }
    for (let i = 0; i < nums.length; i++) {
      if (!visited.has(i)) {
        visited.add(i);
        curr.push(nums[i]);
        dfs(i, curr);
        curr.pop();
        visited.delete(i);
      }
    }
  };
  dfs(0);
  return answer;
};
